daffodil-core/src/test/scala/org/apache/daffodil/dsom/walker/TestDSOMWalker.scala

```diff
@@ -0,0 +1,229 @@
+    val nodeStack: List[Either[TermView, TypeView]] = walker.nodeArr.toList
+    assertEquals(s"Node Stack $nodeStack did not have the expected number of elements", 7, nodeStack.size)
+    nodeStack.head match {
+      case Left(personElement: RootView) =>
+        assertEquals("The root element should be named 'PersonData'", "PersonData", personElement.name)
+      case Left(someClass) => fail(s"The Root element was not of type RootView, it was a ${someClass.getClass}!")
+      case Right(someClass) => fail(s"The Root element was not of type RootView, it was a ${someClass.getClass}!")
+    }
+    nodeStack(1) match {
+      case Right(_: ComplexTypeView) =>
+      case Left(someClass) => fail(s"The Root element did not contain a complexType wrapper child, it was a ${someClass.getClass}!")
+      case Right(someClass) => fail(s"The Root element element did not contain a complexType wrapper child, it was a ${someClass.getClass}!")
+    }
+    nodeStack(2) match {
+      case Left(_: SequenceView) =>
+      case Left(someClass) => fail(s"The complexType element did not contain a Sequence child, it was a ${someClass.getClass}!")
+      case Right(someClass) => fail(s"The complexType element did not contain a Sequence child, it was a ${someClass.getClass}!")
+    }
+    nodeStack(3) match {
+      case Left(_: ChoiceView) =>
+      case Left(someClass) => fail(s"The Sequence element did not contain a Choice child, it was a ${someClass.getClass}!")
+      case Right(someClass) => fail(s"The Sequence element did not contain a Choice child, it was a ${someClass.getClass}!")
+    }
+    nodeStack(4) match {
+      case Left(nameElement: ElementBaseView) =>
+        assertEquals("The second child element should be named 'age'", "age", nameElement.name)
+        assertEquals("The second child element should be a simple type", true, nameElement.isSimpleType)
+      case Left(someClass) => fail(s"The Sequence element did not contain a second Element child, it was a ${someClass.getClass}!")
+      case Right(someClass) => fail(s"The Sequence element did not contain a second Element child, it was a ${someClass.getClass}!")
+    }
+    nodeStack(5) match {
+      case Right(simpleType: SimpleTypeView) =>
+        simpleType.primType match {
+          case _:IntView =>
+          case _ => fail(s"The 'age' element should be of simple type Int")
+        }
+      case Left(someClass) => fail(s"The 'age' element did not have a SimpleTypeView, it was a ${someClass.getClass}!")
+      case Right(someClass) => fail(s"The 'age' element did not have a SimpleTypeView, it was a ${someClass.getClass}!")
+    }
+    nodeStack(6) match {
+      case Left(_: GroupRefView) =>
+      case Left(someClass) => fail(s"The Sequence element did not contain a GroupRef child, it was a ${someClass.getClass}!")
+      case Right(someClass) => fail(s"The Sequence element did not contain a GroupRef child, it was a ${someClass.getClass}!")
+    }
+  }
```

Comment:
These feels a little verbose to me. All the fail messages are perhaps helpful when a test fails, but that should be rare, and it makes it difficult to visually determine that the tests are checking the right things, which I'd argue is more important.

I think part of the problem this looks so verbose is the the node stack being an Either, so you need to do a bunch of left/right checks. But that doesn't give us much. You still need to do type matcing inside the Left/Right, so the Left/Right just adds a layer of indirection. Instead, you can just make nodeArray be an Array[AnyRef].

Then, if we removed all the matching and messages, the above can be condensed down to something like this:
```scala
assertEquals(7, nodeStack.size)
assertTrue(nodeStack(0).isInstanceOf[RootView])
assertEquals(nodeStack(0).asInstanceOf[RootView].name, "PersonData")
assertTrue(nodeStack(1).isInstanceOf[ComplexTypeView])
assertTrue(nodeStack(2).isInstanceOf[SequenceView])
assertTrue(nodeStack(3).isInstanceOf[ChoiceView])
assertTrue(nodeStack(4).isInstanceOf[ElementBaseView])
assertEquals(nodeStack(4).asInstanceOf[ElementBaseView].name, "age")
assertEquals(nodeStack(4).asInstanceOf[ElementBaseView].isSimpleType, true)
assertTrue(nodeStack(5).isInstanceOf[SimpleTypeView])
assertTrue(nodeStack(5).asInstanceOf[SimpleTypeView].primType.isInstanceOf[IntView])
assertTrue(nodeStack(6).isInstanceOf[GroupRefView])
```
Which I think is much easier to visually confirm that we're checking the right things. The suggestion I previously made about having concrete implemenations of the Views and an equal function make these checks a bit more automated and maybe a bit more clear (probably gets rid of all the instance of stuff), at the expense of needing a bit more code. But something more along the lines of this doesn't seem unreasonable to me.

Note that this is much less scala-y, but readability often trumps that, especially when it comes to tests.

Reply:
I agree that it is robust, and I had it this way mainly for the sake of pattern matching and being able to figure out types of specific View elements on the stack when tests were failing.  That being said, for simplicity sake, I think the series of asserts based on `isInstanceOf` and `asInstanceOf` would be best.  I tried implementing a case class as you suggested above, and it turned out to be an even longer solution.